Make collection caching explicit.

kaspth opened the pull request:
Having collection caching that wraps templates and automatically tries
to infer if they are cachable proved to be too much of a hassle.

We'd rather have it be something you explicitly turn on.

This removes much of the code and docs to explain the previous automatic
behavior.

This change also removes scoped cache keys and passing cache_options.

cc @dhh

kaspth committed:
Make collection caching explicit.

Having collection caching that wraps templates and automatically tries
to infer if they are cachable proved to be too much of a hassle.

We'd rather have it be something you explicitly turn on.

This removes much of the code and docs to explain the previous automatic
behavior.

This change also removes scoped cache keys and passing cache_options.

kaspth commented on actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb:
```diff
@@ -53,10 +37,13 @@ def expanded_cache_key(key)
         key.frozen? ? key.dup : key # #read_multi & #write may require mutability, Dalli 2.6.0.
```
Check out the explicit `@template` use just above to extract a `virtual_path` because we reference this, we can't support collections rendering multiple templates. Well, we can, there's just no telling that the collection cache reuses the individual fragment caching.

I don't know of a better way to get a hold of the `virtual_path` from here than accessing it through the templates. This likely means we'll have to split switch to eager-like loading of the templates as done here: https://github.com/rails/rails/pull/19577/files#diff-738cc3c10058587e78559e48066ad0bfR58 (another stab at supporting multiple templates).

kaspth replied:
But anyway, that shouldn't impede reviews on this.

dhh commented:
Great to see all this code cut out. I'd like to see better logging as part of this. When you do a render/collection call, that should have a summary line ala: `Rendered collection of chats/transcripts/_line.html.erb [5 / 7 cache hits] (2.9ms)`. This is in addition to the individual lines of rendering templates.

kaspth commented:
@dhh wouldn't you want to know which of the templates missed the cache?

How much sense does the read_multi instrumentation + logging make with this additional logging?

dhh commented:
What's currently happening with the "read_multi instrumentation + logging"?
I don't really see anything?

Which templates? Isn't it always the same template if you're rendering a
collection? Wouldn't need to know which IDs of the collection that missed,
just the count imo.

On Tue, Feb 16, 2016 at 5:32 PM, Kasper Timm Hansen <
notifications@github.com> wrote:

> @dhh <https://github.com/dhh> wouldn't you want to know which of the
> templates missed the cache?
>
> How much sense does the read_multi instrumentation + logging make with
> this additional logging?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/23695#issuecomment-184755600>.
>

kaspth commented:
I guess you not seeing anything logged is another argument for making this explicit :D

Here's what I see when I render a collection with:

```erb
<%= render partial: 'notifications/notification', collection: @notifications, cached: true %>
```

```
Caches multi read:
- views/notifications/1-20160104214800726582/94ec7ce9b34acbe92cfdb2662cded66a
- views/notifications/2-20160104214800726582/94ec7ce9b34acbe92cfdb2662cded66a
- views/notifications/3-20160104214800726582/94ec7ce9b34acbe92cfdb2662cded66a
- views/notifications/4-20160104214800726582/94ec7ce9b34acbe92cfdb2662cded66a
- views/notifications/5-20160104214800726582/94ec7ce9b34acbe92cfdb2662cded66a
- views/notifications/6-20160104214800726582/94ec7ce9b34acbe92cfdb2662cded66a
- views/notifications/7-20160104214800726582/94ec7ce9b34acbe92cfdb2662cded66a
- views/notifications/8-20160104214800726582/94ec7ce9b34acbe92cfdb2662cded66a
- views/notifications/9-20160104214800726582/94ec7ce9b34acbe92cfdb2662cded66a
```

> Which templates? Isn't it always the same template if you're rendering a
> collection? Wouldn't need to know which IDs of the collection that missed,
> just the count imo.

Tangent point: it's not always the same template in, say, an STI setup. But yeah, hit rate should be plenty.

kaspth commented:
Realized we don't honor `config.action_controller.perform_caching`. Doubly odd is Action View's `cache` method checking the controller. Think it signifies we need a shared `perform_caching` option for Action View and Action Controller. Maybe `perform_caching` is a little too general for the top level `config`. Will think about this some more. Suggestions welcome 😁

kaspth commented:
Drafted up an implementation of your logging suggestions. It's way easier to see what's going on for sure:

```
97 other lines snipped...
- views/notifications/98-20160218204758471328/94ec7ce9b34acbe92cfdb2662cded66a
- views/notifications/99-20160218204758471328/94ec7ce9b34acbe92cfdb2662cded66a
- views/notifications/100-20160218204758471328/94ec7ce9b34acbe92cfdb2662cded66a
  Rendered collection of notifications/_notification.html.erb [0 / 100 cache hits] (3396.5ms)
Cache write: /Users/kasperhansen/Documents/code/collection_caching_test/tmp/cache/7CD/611/views%2Fnotifications%2F1-20160218204758471328%2F94ec7ce9b34acbe92cfdb2662cded66a
Cache write: /Users/kasperhansen/Documents/code/collection_caching_test/tmp/cache/7CE/9A1/views%2Fnotifications%2F2-20160218204758471328%2F94ec7ce9b34acbe92cfdb2662cded66a
More lines snipped...
```

I'm leaning towards just yanking that whole `read_multi` instrumentation.

kaspth committed:
Instrument cached collection renders.

Augments the collection caching with some instrumentation that's logged.

For collections that have been cached like:

```ruby
<%= render partial: 'notifications/notification', collection: @notifications, cached: true %>
```

We'll output a line showing how many cache hits we had when rendering it:

```
  Rendered collection of notifications/_notification.html.erb [0 / 100 cache hits] (3396.5ms)
```

dhh commented:
Agree we can drop the other instrumentation. This seems better. I say we merge and feel it out.

rafaelfranca commented on actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb:
```diff
@@ -12,39 +10,29 @@ module CollectionCaching # :nodoc:
+        instrument(:cached_collection,
+          count: keyed_collection.size, hits: cached_partials.size,
+          identifier: @template.identifier || @path || 'cached_collection') do
+          rendered_partials = @collection.empty? ? [] : yield
```
This indentation is weird. Better to keep all the arguments in the same line.

sgrif replied:
Could also split them up into one argument per line.

```ruby
instrument(
  :cached_collection,
  count: keyed_collection.size,
  hits: cached_partials.size,
  identifier: @template.identifier || @path || 'cached_collection',
) do
  # ...
end
```

rafaelfranca commented:
LGTM

kaspth committed:
Revert "Instrument read_multi".

Reevaluating the log output generated from this instrumentation,
we've found that it wasn't all that useful in practice.

```
Caches multi read:
- views/david/2/4184ab71db6849621a4d8820fcd2c0ad
- views/david/2/4184ab71db6849621a4d8820fcd2c0ad
- views/david/3/4184ab71db6849621a4d8820fcd2c0ad
- views/david/3/4184ab71db6849621a4d8820fcd2c0ad
```

If rendering many templates the output is inscrutable, and it's impossible
to see how many cache misses there were.

Revert ca6aba7f30 and implement a better way later.

kaspth commented:
Thanks for the reviews 🖖

ghiculescu commented on actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb:
```diff
@@ -11,40 +9,25 @@ module CollectionCaching # :nodoc:
-      def cache_collection_render
-        return yield unless cache_collection?
+      def cache_collection_render(instrumentation_payload)
+        return yield unless @options[:cached]
```
out of curiosity, what was the rationale of going from `@options[:cache]` to `@options[:cached]`?

this bit us when we upgraded to rails 5 and removed `multi_fetch_fragments` from our Gemfile - it took ages to figure out why collection caching just wasn't happening anymore.
